libopenage/gamestate/score.cpp

```diff
@@ -0,0 +1,111 @@
+// Copyright 2017-2017 the openage authors. See copying.md for legal info.
+
+#include <cmath>
+
+#include "player.h"
+#include "score.h"
+#include "team.h"
+#include "../log/log.h"
+
+namespace openage {
+
+Score::Score()
+	:
+	score{0},
+	score_total{0},
+	score_exploration{0},
+	score_resources{0} {
+}
+
+void Score::add_score(const score_category cat, double value) {
+	this->add_score(cat, (int) std::lround(value));
+}
+
+void Score::add_score(const score_category cat, int value) {
+	this->score[(int) cat] += value;
+	this->update_score();
+}
+
+void Score::remove_score(const score_category cat, double value) {
+	this->remove_score(cat, (int) std::lround(value));
+}
+
+void Score::remove_score(const score_category cat, int value) {
+	this->score[(int) cat] -= value;
+	this->update_score();
+}
+
+void Score::update_map_explored(double progress) {
+	this->remove_score(score_category::technology, this->score_exploration);
+	this->score_exploration = progress * 1000;
+	this->add_score(score_category::technology, this->score_exploration);
+}
+
+void Score::update_resources(const ResourceBundle & resources) {
+	this->remove_score(score_category::economy, this->score_resources);
+	this->score_resources = resources.sum() * 0.1;
+	this->add_score(score_category::economy, this->score_resources);
+}
+
+void Score::update_score() {
+	this->score_total = 0;
+	for (int i=0; i<(int) score_category::SCORE_CATEGORY_COUNT; i++) {
+		this->score_total += this->getScore(i);
+	}
+}
+
+ScorePlayer::ScorePlayer(Player *player)
+	:
+	Score(),
+	player{player} {
+}
+
+void ScorePlayer::update_score() {
+	Score::update_score();
+	// update team score
+	// TODO fix player->team should be nullptr
+	if (player->team) {
+		// TODO rework team update
```

Comment:
The problem here is that the players  are passed to the Team as consts. One solution could be to have the players in Team not const